managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2357,6 +2357,42 @@ private void maybeOffloadInBackground(CompletableFuture<PositionImpl> promise) {
         }
     }
 
+    // Although we have caught the connection loss exception on the meta store, to avoid other exceptions cause
+    // the mismatch between meta store and in memory, we refresh the ledger info list when the offload execute
+    // failed by badversion
+    private void asyncRefreshLedgersInfoOnBadVersion(ManagedLedgerException exception) {
+        if (!(exception instanceof BadVersionException)) {
+            return;
+        }
+        if (!metadataMutex.tryLock()) {
+            scheduledExecutor.schedule(
+                () -> asyncRefreshLedgersInfoOnBadVersion(exception), 100, TimeUnit.MILLISECONDS);
+            return;
+        }
+        store.getManagedLedgerInfo(name, false, new MetaStoreCallback<>() {
+            @Override
+            public void operationComplete(ManagedLedgerInfo mlInfo, Stat stat) {
+                ledgersStat = stat;
+                try {
+                    synchronized (ManagedLedgerImpl.this) {
```

Comment:
I think here should be an issue because maybe there is another operation that updated the ledger list (take the lock first), and then here takes the lock which will mess up the ledger list again.

Reply:
All the ledger's update operations should be guarded by the metadata lock because it needs to make sure the ledger stat is the latest version. 
And the synchronized to make sure there hasn't remove/add operation when the ledger is closing or creating.

I haven't found other places to operate the map without locks. Do you know if other places still have concurrency issues?

Reply:
> And the synchronized to make sure there hasn't remove/add operation when the ledger is closing or creating.

Yes, that is what I want to say. If there are other operations that will update the `ledgers`, it will introduce the problem.

For example:

1. asyncRefreshLedgersInfoOnBadVersion get the metadataMutex lock
2. we got the returned `ManageLedgerInfo` at line 2370
3. another operation changed the `ledgers`,  we will waiting here line 2373
4. after the `ledgers` changed, we will continue to run lines 2374 - 2377
5. we will lose the changes at step 3  

And it also might introduce deadlock? someone is getting the `synchronized` but waiting for the `metadataMutex` lock, here is waiting for the `synchronized` lock but can't release the `metadataMutex` lock

Reply:
The method `synchronized void ledgerClosed(final LedgerHandle lh)` will update the `ledgers` only with `synchronized` lock, no `metadataMutex`

Reply:
Because when we use metadataMutex, we are trying to lock it, I think it won't cause a deadlock?

When we lock the metadataMutex, I saw all other places will retry. It won't update the ledgers or metadata successfully. Then it has to wait for the refresh done and then do other things.